src/image_builder.py

```diff
@@ -154,109 +156,198 @@ def image_builder(buildspec):
+            FORMATTER.print("No errors")
+    return is_any_build_failed, is_any_build_failed_size_limit
+
+def upload_metrics(images, BUILDSPEC, is_any_build_failed, is_any_build_failed_size_limit):
+
+    FORMATTER.title("Uploading Metrics")
+    is_any_build_failed = False
+    is_any_build_failed_size_limit = False
+    metrics = Metrics(
+        context=constants.BUILD_CONTEXT,
+        region=BUILDSPEC["region"],
+        namespace=constants.METRICS_NAMESPACE,
+    )
+    for image in images:
+        try:
+            metrics.push_image_metrics(image)
+        except Exception as e:
+            if is_any_build_failed or is_any_build_failed_size_limit:
+                raise Exception(f"Build failed.{e}")
             else:
-                FORMATTER.print("No errors")
+                raise Exception(f"Build passed. {e}")
 
-        FORMATTER.title("Uploading Metrics")
-        metrics = Metrics(
-            context=constants.BUILD_CONTEXT,
-            region=BUILDSPEC["region"],
-            namespace=constants.METRICS_NAMESPACE,
-        )
-        for image in IMAGES:
-            try:
-                metrics.push_image_metrics(image)
-            except Exception as e:
-                if is_any_build_failed or is_any_build_failed_size_limit:
-                    raise Exception(f"Build failed.{e}")
-                else:
-                    raise Exception(f"Build passed. {e}")
+    if is_any_build_failed_size_limit:
+        raise Exception("Build failed because of file limit")
 
-        if is_any_build_failed_size_limit:
-            raise Exception("Build failed because of file limit")
+    FORMATTER.separator()
 
-        FORMATTER.separator()
+def build_images(images, make_dummy_boto_client=False):
+    THREADS = {}
+    # In the context of the ThreadPoolExecutor each instance of image.build submitted
+    # to it is executed concurrently in a separate thread.
+    with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
+        #### TODO: Remove this entire if block when https://github.com/boto/boto3/issues/1592 is resolved ####
+        if make_dummy_boto_client: 
+            get_dummy_boto_client()
+        for image in images:
+            FORMATTER.print(f"image_object.context {image.context}")
+            THREADS[image.name] = executor.submit(image.build)
+    # the FORMATTER.progress(THREADS) function call also waits until all threads have completed
+    FORMATTER.progress(THREADS)
+
+#### TODO: Remove this entire method when https://github.com/boto/boto3/issues/1592 is resolved ####
+def get_dummy_boto_client():
+    # In absence of this method, the behaviour documented in https://github.com/boto/boto3/issues/1592 is observed.
+    # If this function is not added, boto3 fails because boto3 sessions are not thread safe.
+    # However, once a dummy client is created, it is ensured that the calls are thread safe.
+    import boto3
```

Comment:
`boto3` is a module that is common enough that we can reasonably expect it to always be available in the environment when performing builds (to be confirmed by checking `src/requirements.txt`.

We can always remove `boto3` later if it is unused. Most Python IDEs highlight unused imports so that developers can easily discover them and decide whether to remove those imports.

Reply:
Done.